### PR TITLE
Support custom catalog for iceberg external table

### LINKED ISSUE #3826: Support Iceberg custom catalog
## Feature request
Support Iceberg custom catalog so that users can build their own catalog to store metadata of Iceberg
related with https://github.com/StarRocks/starrocks/issues/1030

### PR DESCRIPTION
**Pull request by mxdzs0612**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3826

## Problem Summary(Required) ：

Add support for custom catalog which can be defined by users themselves when creating iceberg external table.

The custom catalog should be in the form of IcebergHiveCatalog, in other words extending BaseMetastoreCatalog and implementing IcebergCatalog. The catalog JAR should be placed into each fe/lib directory, and FE has to be restarted before custom catalog works.

Usage:
```sql
CREATE EXTERNAL RESOURCE "iceberg0"
PROPERTIES (
  "type" = "iceberg",
  "starrocks.catalog-type"="CUSTOM",
  "iceberg.catalog-impl"="{The full class name of custom catalog}"
);
```
Extra config users defined can be added in table properties when executing CREATE EXTERNAL TABLE, see #2225.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 14 / 32 (43.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|1|8|12.50%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|1|5|20.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|3|6|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 19 / 32 (59.38%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|3|6|50.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|4|8|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 21 / 36 (58.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|4|8|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Review by caneGuy — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java [hunk 1]
```diff
@@ -44,6 +44,7 @@ public class IcebergTable extends Table {
 
     private static final String ICEBERG_CATALOG = "starrocks.catalog-type";
     private static final String ICEBERG_METASTORE_URIS = "iceberg.catalog.hive.metastore.uris";
+    private static final String ICEBERG_IMPL = "iceberg.catalog-impl";
```
Comment (caneGuy) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:47]: will this compatible with old version？

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java [hunk 2]
```diff
@@ -33,13 +33,17 @@ public class IcebergResource extends Resource {
...
     private static final String ICEBERG_METASTORE_URIS = "iceberg.catalog.hive.metastore.uris";
+    private static final String ICEBERG_IMPL = "iceberg.catalog-impl";
 
     @SerializedName(value = "catalogType")
     private String catalogType;
 
     @SerializedName(value = "metastoreURIs")
     private String metastoreURIs;
 
+    @SerializedName(value = "impl")
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java:44]: I think should give more clearer name.
like `customImpl`

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java [hunk 3]
```diff
@@ -144,12 +153,20 @@ private void validate(Map<String, String> properties) throws DdlException {
             case HIVE_CATALOG:
                 icebergProperties.put(ICEBERG_METASTORE_URIS, icebergResource.getHiveMetastoreURIs());
                 break;
+            case CUSTOM_CATALOG:
+                icebergProperties.put(ICEBERG_IMPL, icebergResource.getIcebergImpl());
+                for (String key : copiedProps.keySet()) {
+                    icebergProperties.put(key, copiedProps.remove(key));
+                }
+                // todo: validate custom catalog
+                // cannot validate columns for custom catalog at present, return directly.
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:162]: why cannot?

**Comment by DorianZheng** [comment]
I have left a comment, rest LGTM.

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java (hunk 1 above)
Comment (mxdzs0612) [reply fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:47]: There is no influence on other kinds of catalogs.

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java [hunk 4]
```diff
@@ -33,13 +33,17 @@ public class IcebergResource extends Resource {
...
     private static final String ICEBERG_METASTORE_URIS = "iceberg.catalog.hive.metastore.uris";
+    private static final String ICEBERG_IMPL = "iceberg.catalog-impl";
 
     @SerializedName(value = "catalogType")
     private String catalogType;
 
     @SerializedName(value = "metastoreURIs")
     private String metastoreURIs;
 
+    @SerializedName(value = "customImpl")
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java:44]: seems CatalogImpl is a better name.
It is the same as ICEBERG

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java [hunk 5]
```diff
@@ -64,6 +68,17 @@ protected void setProperties(Map<String, String> properties) throws DdlException
                     throw new DdlException(ICEBERG_METASTORE_URIS + " must be set in properties");
                 }
                 break;
+            case CUSTOM_CATALOG:
+                catalogImpl = properties.get(ICEBERG_IMPL);
+                if (StringUtils.isBlank(catalogImpl)) {
+                    throw new DdlException(ICEBERG_IMPL + " must be set in properties");
+                }
+                try {
+                    Thread.currentThread().getContextClassLoader().loadClass(catalogImpl);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java:77]: Am I missing some logic? I think loadClass only occurs when the resource is created. But when the fe process is restarted, the custom class doesn't seem to be loaded into the jvm.

Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java:77]: This is just to verify whether the jvm has loaded the custom class？ If fe restarts and the loadClass() method will not be called, Will it affect the loadCatalog later?

Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/catalog/IcebergResource.java:77]: Are there other considerations here? In fact, there is no need to restart fe by dynamically loading the custom jar.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 25 / 41 (60.98%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|8|13|61.54%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 25 / 41 (60.98%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|8|13|61.54%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Comment by mxdzs0612** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 25 / 41 (60.98%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|8|13|61.54%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Review by caneGuy — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java [hunk 6]
```diff
@@ -143,13 +152,26 @@ private void validate(Map<String, String> properties) throws DdlException {
...
             case HIVE_CATALOG:
                 icebergProperties.put(ICEBERG_METASTORE_URIS, icebergResource.getHiveMetastoreURIs());
+                validateColumn(IcebergUtil.getIcebergHiveCatalog(icebergResource.getHiveMetastoreURIs()));
+                break;
+            case CUSTOM_CATALOG:
+                icebergProperties.put(ICEBERG_IMPL, icebergResource.getIcebergImpl());
+                for (String key : copiedProps.keySet()) {
+                    icebergProperties.put(key, copiedProps.remove(key));
+                }
+                validateColumn(IcebergUtil.getIcebergCustomCatalog(icebergResource.getIcebergImpl(), icebergProperties));
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:162]: Is it better to move `validateColumn` out of switch-case

**Review by DorianZheng — DISMISSED** [review]
I have left one comment, rest LGTM

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 26 / 42 (61.90%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergResource.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergUtil.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|9|14|64.29%|
|:large_blue_circle:|com/starrocks/external/iceberg/CatalogLoader.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/external/iceberg/IcebergCatalogType.java|1|1|100.00%|

**Review by caneGuy — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**